subprojects/platform-native/src/integTest/groovy/org/gradle/nativeplatform/tasks/GenerateModuleMapIntegrationTest.groovy

```diff
@@ -0,0 +1,105 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.nativeplatform.tasks
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+
+import static org.gradle.util.TextUtil.normaliseLineSeparators
+
+class GenerateModuleMapIntegrationTest extends AbstractIntegrationSpec {
+
+    def "generates a simple module map file"() {
+        buildFile << """
+            import ${GenerateModuleMap.class.canonicalName}
+
+            task generate(type: GenerateModuleMap) {
+                moduleName.set('foo') 
+                moduleMapFile.set(project.layout.buildDirectory.file('moduleMaps/module.modulemap'))
+                publicHeaderDirs.add(file('headers'))
+                publicHeaderDirs.add(file('moreHeaders'))
+                publicHeaderDirs.add(file('does-not-exist'))
+            }
+        """
+        file('headers').createDir()
+        file('moreHeaders').createDir()
+
+        expect:
+        succeeds "generate"
+        def moduleMapFile = file('build/moduleMaps/module.modulemap')
+        moduleMapFile.exists()
+        normaliseLineSeparators(moduleMapFile.text) == content('foo', 'headers', 'moreHeaders')
+    }
+
+    def "generating a module map file is incremental"() {
+        def moduleMapFile = file('build/moduleMaps/module.modulemap')
+        buildFile << """
+            import ${GenerateModuleMap.class.canonicalName}
+
+            task generate(type: GenerateModuleMap) {
+                moduleName.set('foo') 
+                moduleMapFile.set(project.layout.buildDirectory.file('moduleMaps/module.modulemap'))
+                publicHeaderDirs.add(file('headers'))
+                publicHeaderDirs.add(file('moreHeaders'))
+            }
+        """
+        file('headers').createDir()
+        file('moreHeaders').createDir()
+
+        when:
+        succeeds "generate"
+
+        then:
+        executedAndNotSkipped ":generate"
+        moduleMapFile.exists()
+        normaliseLineSeparators(moduleMapFile.text) == content('foo', 'headers', 'moreHeaders')
+
+        when:
+        succeeds "generate"
+
+        then:
+        skipped ":generate"
+
+        when:
+        buildFile << """
+            generate.moduleName.set('bar')
+        """
+        succeeds "generate"
+
+        then:
+        executedAndNotSkipped ":generate"
+        normaliseLineSeparators(moduleMapFile.text) == content('bar', 'headers', 'moreHeaders')
+
+        when:
+        buildFile << """
+            generate.publicHeaderDirs.add(file('yet/more/headers'))
+        """
+        file('yet/more/headers').createDir()
+        succeeds "generate"
+
+        then:
+        executedAndNotSkipped ":generate"
+        normaliseLineSeparators(moduleMapFile.text) == content('bar', 'headers', 'moreHeaders', 'yet/more/headers')
+    }
+
+    String content(String moduleName, String... headerPaths) {
```

Comment:
I think I would prefer if the tests here just had the modulemap contents inline vs recreating the modulemap in the test code.

E.g., the first test would just be something like:
```
normaliseLineSeparators(moduleMapFile.text) == """
module foo {
    umbrella "${file("headers").absolutePath}"
    umbrella "${file("moreHeaders").absolutePath}"
    export *
}
"""
```

Otherwise, it's hard to know what these things are supposed to look like.

Reply:
Fair enough.  Done.